File: typescript/tests/verification/name-coloured-encoder.test.ts
import { describe, it, expect } from 'vitest';
import { PetriNet } from '../../src/core/petri-net.js';
import { Transition } from '../../src/core/transition.js';
import { place } from '../../src/core/place.js';
import { one } from '../../src/core/in.js';
import { outPlace, andPlaces, xorPlaces } from '../../src/core/out.js';
import { matchSpec, matchKey } from '../../src/core/match-spec.js';
import { nameId } from '../../src/core/name.js';
import { MarkingState } from '../../src/verification/marking-state.js';
import { flatten } from '../../src/verification/encoding/net-flattener.js';
import { computePSemiflows } from '../../src/verification/invariant/p-invariant-computer.js';
import { IncidenceMatrix } from '../../src/verification/encoding/incidence-matrix.js';
import { buildColouredPlan } from '../../src/verification/z3/name-coloured-encoder.js';
import type { FragmentMode } from '../../src/verification/analysis/name-fragment.js';

// Z3-free conformance for the name-coloured fragment gate (buildColouredPlan).
// The colour-slot bound `k` comes from a non-negative P-semiflow that weights every
// coloured place: a net whose coloured set has no covering semiflow (an unbounded
// colour leak) falls back to the sound over-approximation; a colour-bounded net is
// admitted for the exact name-coloured path.
describe('name-coloured fragment gate (colour-slot bound)', () => {
  // A budget-bounded mint→join net (same-mint scatter-gather): one mint consumes
  // 1 budget and stamps the fresh colour into both correlated inputs; the join
  // refunds 1 budget (conserving) or refunds an EXTRA token to a non-minting place
  // (budget2) — either way the MINTING budget (budget1) stays conserved, so at most
  // one colour is ever live and the P-semiflow bound admits it.
  function mintJoinNet(extraRefund: boolean) {
    const budget1 = place('budget1');
    const budget2 = place('budget2');
    const a = place<string>('branchA');
    const b = place<string>('branchB');

    const mint = Transition.builder('mint')
      .inputs(one(budget1))
      .outputs(andPlaces(a, b))
      .build();
    const joinOut = extraRefund ? andPlaces(budget1, budget2) : outPlace(budget1);
    const join = Transition.builder('join')
      .inputs(one(a), one(b))
      .match(matchSpec(
        matchKey(a, (s: string) => nameId(s)),
        matchKey(b, (s: string) => nameId(s)),
      ))
      .outputs(joinOut)
      .build();
    const net = PetriNet.builder('mintJoin').transitions(mint, join).build();
    return { net, budget1 };
  }

  function planFor(extraRefund: boolean) {
    const { net, budget1 } = mintJoinNet(extraRefund);
    const flat = flatten(net);
    const initial = MarkingState.builder().tokens(budget1, 1).build();
    const semiflows = computePSemiflows(IncidenceMatrix.from(flat), flat, initial);
    return buildColouredPlan(net, flat, initial, new Set(['budget1', 'budget2']), 'base', new Set(), semiflows);
  }

  it('takes the exact path when the join conserves budget', () => {
    // Refund (1) == mint cost (1): live names ≤ k.
    expect(planFor(false)).not.toBeNull();
  });

  it('admits a join that refunds an extra token to a non-minting place', () => {
    // [NU-053] The extra refund lands in budget2 (a non-minting place), so the minting
    // budget stays conserved and at most one colour is live — the net is colour-bounded
    // and the P-semiflow bound admits it. (The old budget-Φ heuristic wrongly rejected
    // any refund exceeding the mint cost; genuine colour leaks — where a co-minted place
    // accumulates distinct colours — are covered by the leaky-carrier fan-out test below,
    // which still falls back.)
    expect(planFor(true)).not.toBeNull();
  });
});

// === NU-053: EXTENDED coloured-consumer fragment + XOR expansion ===
describe('name-coloured fragment gate (NU-053 EXTENDED + XOR)', () => {
  // Shared driver: flatten, seed budget1=1, and run the gate under `mode`/`carriers`.
  function planFor(net: PetriNet, mode: FragmentMode, carriers: string[]) {
    const flat = flatten(net);
    const budget1 = place('budget1');
    const initial = MarkingState.builder().tokens(budget1, 1).build();
    const semiflows = computePSemiflows(IncidenceMatrix.from(flat), flat, initial);
    return buildColouredPlan(
      net, flat, initial, new Set(['budget1', 'budget2']), mode, new Set(carriers), semiflows,
    );
  }

  // mint→join plus an EXTENDED coloured drain: a non-match transition that consumes
  // one correlated input `a` (count 1) into a plain sink. Rejected under BASE,
  // admitted under EXTENDED.
  function mintJoinDrainNet() {
    const budget1 = place('budget1');
    const a = place<string>('branchA');
    const b = place<string>('branchB');
    const sink = place<string>('sink');

    const mint = Transition.builder('mint').inputs(one(budget1)).outputs(andPlaces(a, b)).build();
    const join = Transition.builder('join')
      .inputs(one(a), one(b))
      .match(matchSpec(matchKey(a, (s: string) => nameId(s)), matchKey(b, (s: string) => nameId(s))))
      .outputs(outPlace(budget1))
      .build();
    const drain = Transition.builder('drain').inputs(one(a)).outputs(outPlace(sink)).build();
    return PetriNet.builder('mintJoinDrain').transitions(mint, join, drain).build();
  }

  // mint→join with an EXTENDED carrier relay: the fork co-mints into `carrier` and
  // join input `b`; the relay threads `carrier`'s name into the other join input
  // `a`. `carrier` is coloured only when declared as a carrier under EXTENDED. When
  // `relayRefunds`, the relay also refunds budget — which must be rejected, since a
  // relay keeps the colour live.
  function mintRelayJoinNet(relayRefunds: boolean) {
    const budget1 = place('budget1');
    const carrier = place<string>('carrier');
    const a = place<string>('branchA');
    const b = place<string>('branchB');

    const mint = Transition.builder('mint').inputs(one(budget1)).outputs(andPlaces(carrier, b)).build();
    const relayOut = relayRefunds ? andPlaces(a, budget1) : outPlace(a);
    const relay = Transition.builder('relay').inputs(one(carrier)).outputs(relayOut).build();
    const join = Transition.builder('join')
      .inputs(one(a), one(b))
      .match(matchSpec(matchKey(a, (s: string) => nameId(s)), matchKey(b, (s: string) => nameId(s))))
      .outputs(outPlace(budget1))
      .build();
    return PetriNet.builder('mintRelayJoin').transitions(mint, relay, join).build();
  }

  // mint→join plus a plain XOR transition (uncoloured), which expands to two flat
  // rows — exercising Part 3 (no 1:1 net↔flat assumption).
  function mintJoinXorNet() {
    const budget1 = place('budget1');
    const a = place<string>('branchA');
    const b = place<string>('branchB');
    const s = place('s');
    const x = place('x');
    const y = place('y');

    const mint = Transition.builder('mint').inputs(one(budget1)).outputs(andPlaces(a, b)).build();
    const join = Transition.builder('join')
      .inputs(one(a), one(b))
      .match(matchSpec(matchKey(a, (v: string) => nameId(v)), matchKey(b, (v: string) => nameId(v))))
      .outputs(outPlace(budget1))
      .build();
    const branch = Transition.builder('branch').inputs(one(s)).outputs(xorPlaces(x, y)).build();
    return PetriNet.builder('mintJoinXor').transitions(mint, join, branch).build();
  }

  // A leaky fork ([NU-053] S2): the mint co-mints its colour into the join inputs
  // `branchA`, `branchB` AND a declared carrier `c`, but the join only re-collects
  // `branchA`, `branchB` and nothing ever consumes `c`. The colour outlives the budget
  // the join refunds, so the real net can hold more than k live colours while the
  // k-colour encoding gets stuck at the freshness guard.
  function mintLeakyCarrierNet() {
    const budget1 = place('budget1');
    const a = place<string>('branchA');
    const b = place<string>('branchB');
    const c = place<string>('c');

    const mint = Transition.builder('mint').inputs(one(budget1)).outputs(andPlaces(a, b, c)).build();
    const join = Transition.builder('join')
      .inputs(one(a), one(b))
      .match(matchSpec(matchKey(a, (s: string) => nameId(s)), matchKey(b, (s: string) => nameId(s))))
      .outputs(outPlace(budget1))
      .build();
    return PetriNet.builder('mintLeakyCarrier').transitions(mint, join).build();
  }

  it('rejects the EXTENDED drain under BASE, admits it under EXTENDED', () => {
    // A non-match consumer of a coloured place is out-of-fragment under BASE and
    // admitted as a drain under EXTENDED.
    const net = mintJoinDrainNet();
    expect(planFor(net, 'base', [])).toBeNull();
    expect(planFor(net, 'extended', [])).not.toBeNull();
  });

  it('admits the carrier relay only under EXTENDED', () => {
    // The carrier place is coloured only when declared under EXTENDED; then the
    // relay threading its name to the join input is an admitted Consume.
    const net = mintRelayJoinNet(false);
    expect(planFor(net, 'base', [])).toBeNull();
    expect(planFor(net, 'extended', ['carrier'])).not.toBeNull();
  });

  it('rejects a relay that also refunds budget', () => {
    // A relay keeps the colour live; if it also refunded budget the freed token
    // could mint a (k+1)-th live colour. The gate must reject it.
    const net = mintRelayJoinNet(true);
    expect(planFor(net, 'extended', ['carrier'])).toBeNull();
  });

  it('rejects a fork whose fan-out is not re-collected by one join', () => {
    // [NU-053] S2: the mint fans its colour into branchA, branchB AND carrier c, but the
    // refunding join only re-collects branchA, branchB — c is never consumed, so the
    // colour outlives its refunded budget and the real net can hold more than k live
    // colours. The gate must reject this (null → sound over-approximation) rather than
    // certify an exact plan the quiescence gate would trust for a false Proven. Contrast
    // the admitted carrier-relay case, where the fork's carrier branch is relayed back
    // into a join input (atomic re-collection).
    expect(planFor(mintLeakyCarrierNet(), 'extended', ['c'])).toBeNull();
  });

  it('no longer blocks the coloured plan on an XOR transition', () => {
    // A plain XOR transition expands to two flat rows; Part 3 drops the old 1:1
    // net↔flat rejection so the mint→join fragment is still recognised.
    const net = mintJoinXorNet();
    expect(planFor(net, 'base', [])).not.toBeNull();
  });
});

File: java/src/test/java/org/libpetri/smt/z3/NameColouredEncoderTest.java
package org.libpetri.smt.z3;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.List;
import java.util.Set;
import java.util.regex.Pattern;

import org.junit.jupiter.api.Test;
import org.libpetri.analysis.EnvironmentAnalysisMode;
import org.libpetri.analysis.FragmentMode;
import org.libpetri.analysis.MarkingState;
import org.libpetri.core.Arc;
import org.libpetri.core.MatchSpec;
import org.libpetri.core.NameId;
import org.libpetri.core.PetriNet;
import org.libpetri.core.Place;
import org.libpetri.core.Transition;
import org.libpetri.smt.SmtProperty;
import org.libpetri.smt.encoding.FlatNet;
import org.libpetri.smt.encoding.IncidenceMatrix;
import org.libpetri.smt.encoding.NetFlattener;
import org.libpetri.smt.invariant.PInvariantComputer;

/**
 * Z3-free conformance for the name-coloured fragment gate
 * ({@link NameColouredEncoder#buildPlan}). Pins the P-semiflow colour-slot bound (a
 * genuine unbounded colour leak must fall back to the sound over-approximation) plus the
 * NU-053 fragment extensions: the EXTENDED coloured-consumer roles (drain / carrier
 * relay), the relay-must-not-refund rule, and XOR-expanded output branches no longer
 * blocking the plan.
 */
class NameColouredEncoderTest {

    /**
     * A budget-bounded mint→join net (same-mint scatter-gather). The mint consumes
     * 1 budget and stamps the fresh colour into both correlated inputs; the join
     * refunds 1 budget (conserving) or 2 (inflating, into a second budget place).
     */
    private static PetriNet mintJoinNet(boolean inflating) {
        var budget1 = Place.of("budget1", Integer.class);
        var budget2 = Place.of("budget2", Integer.class);
        var a = Place.of("branchA", String.class);
        var b = Place.of("branchB", String.class);

        var mint = Transition.builder("mint")
            .inputs(Arc.In.one(budget1))
            .outputs(Arc.Out.and(a, b))
            .build();
        Arc.Out joinOut = inflating ? Arc.Out.and(budget1, budget2) : Arc.Out.place(budget1);
        var join = Transition.builder("join")
            .inputs(Arc.In.one(a), Arc.In.one(b))
            .match(MatchSpec.builder()
                .key(a, (String s) -> NameId.of(s))
                .key(b, (String s) -> NameId.of(s))
                .build())
            .outputs(joinOut)
            .build();
        return PetriNet.builder("mintJoin").transitions(mint, join).build();
    }

    /**
     * A mint→join net plus an EXTENDED coloured drain: a non-match transition that
     * consumes one correlated input {@code a} (count 1) into a plain sink. Rejected under
     * BASE (a non-match consumer of a coloured place), admitted under EXTENDED.
     */
    private static PetriNet mintJoinDrainNet() {
        var budget1 = Place.of("budget1", Integer.class);
        var a = Place.of("a", String.class);
        var b = Place.of("b", String.class);
        var sink = Place.of("sink", String.class);

        var mint = Transition.builder("mint")
            .inputs(Arc.In.one(budget1))
            .outputs(Arc.Out.and(a, b))
            .build();
        var join = Transition.builder("join")
            .inputs(Arc.In.one(a), Arc.In.one(b))
            .match(MatchSpec.builder()
                .key(a, (String s) -> NameId.of(s))
                .key(b, (String s) -> NameId.of(s))
                .build())
            .outputs(Arc.Out.place(budget1))
            .build();
        var drain = Transition.builder("drain")
            .inputs(Arc.In.one(a))
            .outputs(Arc.Out.place(sink))
            .build();
        return PetriNet.builder("mintJoinDrain").transitions(mint, join, drain).build();
    }

    /**
     * A mint→join net with an EXTENDED carrier relay: the fork co-mints into the carrier
     * place {@code carrier} and the join input {@code b}; a relay threads {@code carrier}'s
     * name into the other join input {@code a}; the join correlates {@code a} and {@code b}.
     * {@code carrier} is coloured only when declared as a carrier place under EXTENDED. When
     * {@code relayRefunds}, the relay also refunds budget — which must be rejected, since a
     * relay keeps the colour live.
     */
    private static PetriNet mintRelayJoinNet(boolean relayRefunds) {
        var budget1 = Place.of("budget1", Integer.class);
        var carrier = Place.of("carrier", String.class);
        var a = Place.of("a", String.class);
        var b = Place.of("b", String.class);

        var mint = Transition.builder("mint")
            .inputs(Arc.In.one(budget1))
            .outputs(Arc.Out.and(carrier, b))
            .build();
        Arc.Out relayOut = relayRefunds ? Arc.Out.and(a, budget1) : Arc.Out.place(a);
        var relay = Transition.builder("relay")
            .inputs(Arc.In.one(carrier))
            .outputs(relayOut)
            .build();
        var join = Transition.builder("join")
            .inputs(Arc.In.one(a), Arc.In.one(b))
            .match(MatchSpec.builder()
                .key(a, (String s) -> NameId.of(s))
                .key(b, (String s) -> NameId.of(s))
                .build())
            .outputs(Arc.Out.place(budget1))
            .build();
        return PetriNet.builder("mintRelayJoin").transitions(mint, relay, join).build();
    }

    /**
     * A mint→join net plus a plain XOR transition (uncoloured), which expands to two flat
     * rows — exercising NU-053 Part 3 (no 1:1 net↔flat assumption).
     */
    private static PetriNet mintJoinXorNet() {
        var budget1 = Place.of("budget1", Integer.class);
        var a = Place.of("a", String.class);
        var b = Place.of("b", String.class);
        var src = Place.of("src", Integer.class);
        var x = Place.of("x", Integer.class);
        var y = Place.of("y", Integer.class);

        var mint = Transition.builder("mint")
            .inputs(Arc.In.one(budget1))
            .outputs(Arc.Out.and(a, b))
            .build();
        var join = Transition.builder("join")
            .inputs(Arc.In.one(a), Arc.In.one(b))
            .match(MatchSpec.builder()
                .key(a, (String s) -> NameId.of(s))
                .key(b, (String s) -> NameId.of(s))
                .build())
            .outputs(Arc.Out.place(budget1))
            .build();
        var branch = Transition.builder("branch")
            .inputs(Arc.In.one(src))
            .outputs(Arc.Out.xor(x, y))
            .build();
        return PetriNet.builder("mintJoinXor").transitions(mint, join, branch).build();
    }

    /**
     * A leaky fork ([NU-053] S2): the mint co-mints its colour into the join inputs
     * {@code a}, {@code b} AND a declared carrier {@code c}, but the join only re-collects
     * {@code a}, {@code b} and nothing ever consumes {@code c}. The colour outlives the
     * budget the join refunds, so the real net can hold more than k live colours while the
     * k-colour encoding gets stuck at the freshness guard — an under-approximation that
     * could report a false PROVEN.
     */
    private static PetriNet mintLeakyCarrierNet() {
        var budget1 = Place.of("budget1", Integer.class);
        var a = Place.of("a", String.class);
        var b = Place.of("b", String.class);
        var c = Place.of("c", String.class);

        var mint = Transition.builder("mint")
            .inputs(Arc.In.one(budget1))
            .outputs(Arc.Out.and(a, b, c))
            .build();
        var join = Transition.builder("join")
            .inputs(Arc.In.one(a), Arc.In.one(b))
            .match(MatchSpec.builder()
                .key(a, (String s) -> NameId.of(s))
                .key(b, (String s) -> NameId.of(s))
                .build())
            .outputs(Arc.Out.place(budget1))
            .build();
        return PetriNet.builder("mintLeakyCarrier").transitions(mint, join).build();
    }

    private static NameColouredEncoder.ColouredPlan planFor(
            PetriNet net, FragmentMode mode, String... carriers) {
        return planFor(net, mode, 1, carriers);
    }

    private static NameColouredEncoder.ColouredPlan planFor(
            PetriNet net, FragmentMode mode, int budgetTokens, String... carriers) {
        var flat = NetFlattener.flatten(net, Set.of(), EnvironmentAnalysisMode.ignore());
        var initial = MarkingState.builder()
            .tokens(Place.of("budget1", Integer.class), budgetTokens)
            .build();
        var matrix = IncidenceMatrix.from(flat);
        var semiflows = PInvariantComputer.computePSemiflows(matrix, flat, initial);
        return NameColouredEncoder.buildPlan(
            net, flat, initial, Set.of("budget1", "budget2"), mode, Set.of(carriers), semiflows);
    }

    @Test
    void zeroBudgetYieldsTheExactZeroSlotPlan() {
        // NU-053 AC6: with no budget token the covering semiflow's initial sum is zero, and
        // k = 0 is an exact plan rather than a fallback — no coloured token can ever exist
        // (Semiflow.lean, vacuous_colour_layer).
        var plan = planFor(mintJoinNet(false), FragmentMode.BASE, 0);
        assertNotNull(plan, "k = 0 is a plan, not a fallback");
        assertEquals(0, plan.k);
    }

    @Test
    void budgetConservingJoinTakesExactPath() {
        // Refund (1) == mint cost (1): live names ≤ k, so the exact name-coloured
        // encoding is used.
        assertNotNull(planFor(mintJoinNet(false), FragmentMode.BASE));
    }

    @Test
    void budgetRefundToNonmintingPlaceStaysBounded() {
        // [NU-053] A join that refunds an extra token to a NON-minting place keeps the
        // minting budget conserved, so at most one colour is live — the net is
        // colour-bounded and the P-semiflow bound admits it. (The old budget-Φ heuristic
        // wrongly rejected any refund exceeding the mint cost; genuine colour leaks — where
        // a co-minted place accumulates distinct colours — are covered by
        // extendedLeakyCarrierFanoutRejected, which still falls back.)
        assertNotNull(planFor(mintJoinNet(true), FragmentMode.BASE));
    }

    @Test
    void extendedDrainRejectedUnderBaseAdmittedUnderExtended() {
        // A non-match consumer of a coloured place is out-of-fragment under BASE and
        // admitted as a drain under EXTENDED.
        var net = mintJoinDrainNet();
        assertNull(planFor(net, FragmentMode.BASE));
        assertNotNull(planFor(net, FragmentMode.EXTENDED));
    }

    @Test
    void extendedCarrierRelayAdmittedOnlyUnderExtended() {
        // The carrier place is coloured only when declared under EXTENDED; then the relay
        // threading its name to the join input is an admitted Consume.
        var net = mintRelayJoinNet(false);
        assertNull(planFor(net, FragmentMode.BASE));
        assertNotNull(planFor(net, FragmentMode.EXTENDED, "carrier"));
    }

    @Test
    void extendedRelayRefundingBudgetRejected() {
        // A relay keeps the colour live; if it also refunded budget the freed token could
        // mint a (k+1)-th live colour. buildPlan must reject it.
        var net = mintRelayJoinNet(true);
        assertNull(planFor(net, FragmentMode.EXTENDED, "carrier"));
    }

    @Test
    void extendedLeakyCarrierFanoutRejected() {
        // [NU-053] S2: the mint fans its colour into a, b AND carrier c, but the refunding
        // join only re-collects a, b — c is never consumed, so the colour outlives its
        // refunded budget and the real net can hold more than k live colours. buildPlan
        // must reject this (null → sound over-approximation) rather than certify an exact
        // plan the quiescence gate would trust for a false Proven. Contrast the admitted
        // carrier-relay case, where the fork's carrier branch is relayed back into a join
        // input (atomic re-collection).
        assertNull(planFor(mintLeakyCarrierNet(), FragmentMode.EXTENDED, "c"));
    }

    @Test
    void xorTransitionNoLongerBlocksThePlan() {
        // A plain XOR transition expands to two flat rows; NU-053 Part 3 drops the old 1:1
        // net↔flat rejection so the mint→join fragment is still recognised.
        assertNotNull(planFor(mintJoinXorNet(), FragmentMode.BASE));
    }

    /**
     * The coloured quiescence arms ([VER-002] DeadlockFree / TerminatesAtSink, [NU-040]
     * JoinedOrDeadLettered) had NO test in any language before the VER-002 split — a
     * drift in them would have passed every gate, because the shared fixtures either
     * route to Route B or use {@code place-bound}. Mirrors Rust's
     * {@code coloured_quiescence_arms_differ_by_property}.
     */
    @Test
    void colouredQuiescenceArmsDifferByProperty() {
        var net = mintJoinNet(false);
        var flat = NetFlattener.flatten(net, Set.of(), EnvironmentAnalysisMode.ignore());
        var initial = MarkingState.builder()
            .tokens(Place.of("budget1", Integer.class), 1)
            .build();
        var plan = planFor(net, FragmentMode.BASE);
        assertNotNull(plan, "mint→join is in-fragment");

        var branchA = Place.of("branchA", String.class);
        var branchB = Place.of("branchB", String.class);
        Set<Place<?>> sinks = Set.of(branchA);

        String tas = violationTerm(
            encodeColoured(plan, flat, initial, SmtProperty.terminatesAtSink(), sinks));
        // Derive the sink-is-empty clause from TerminatesAtSink's own output rather than
        // reaching into the encoder's private aggregate() helper: it is the clause the
        // arm appends last, over branchA's colour slots.
        var m = Pattern.compile("\\(= \\(\\+[^()]*\\) 0\\)").matcher(tas);
        assertTrue(m.find(), () -> "TerminatesAtSink must emit a sink-is-empty term:\n" + tas);
        String sinkIsEmpty = m.group();
        assertTrue(tas.endsWith(sinkIsEmpty + ")"),
            () -> "the sink clause is the arm's last conjunct:\n" + tas);

        String dl = violationTerm(
            encodeColoured(plan, flat, initial, SmtProperty.deadlockFree(), sinks));
        assertNotEquals(tas, dl, "the two VER-002 properties must encode differently");
        assertFalse(dl.contains(sinkIsEmpty),
            () -> "strict DeadlockFree must not require the sink to be empty:\n" + dl);
        assertTrue(dl.contains("(or (>= "),
            () -> "DeadlockFree must offer a stranded-token disjunction:\n" + dl);

        String jdl = violationTerm(encodeColoured(
            plan, flat, initial, SmtProperty.joinedOrDeadLettered(branchB), sinks));
        assertNotEquals(dl, jdl, "the NU-040 arm must not reuse the DeadlockFree clause");
        assertFalse(jdl.contains(sinkIsEmpty),
            () -> "NU-040 AC4: a declared sink must not excuse a stranded group:\n" + jdl);
    }

    private static String encodeColoured(
            NameColouredEncoder.ColouredPlan plan, FlatNet flat, MarkingState initial,
            SmtProperty property, Set<Place<?>> sinks) {
        var encoding = NameColouredEncoder.encode(plan, flat, initial, property, List.of(), sinks);
        assertNotNull(encoding, "in-fragment net encodes");
        return encoding.smt2();
    }

    /** The {@code Bad(M)} term of the script's error rule, without the Reachable guard. */
    private static String violationTerm(String smt2) {
        int error = smt2.indexOf(")\n      Error)))");
        assertTrue(error > 0, "the script carries an error rule");
        int reachable = smt2.lastIndexOf("(Reachable ", error);
        return smt2.substring(smt2.indexOf(") ", reachable) + 2, error);
    }
}
